src/explorer/Explorers/Components/ExplorerComponent.cs

```diff
@@ -0,0 +1,35 @@
+namespace Explorer.Explorers.Components
+{
+    using System.Threading.Tasks;
+
+    using Diffix;
+    using Explorer.Common;
+
+    internal abstract class ExplorerComponent<TResult>
```

Comment:
Yes, the distinction between `ExplorerBase`, `ExplorerComponent` and even `Exploration` is not super clear, I agree. 

I think some parts of `ExplorerBase` can be merged into `Exploration` and others into `ExplorerComponent`, there's no need to have three distinct classes. However for now I wanted to maintain 'backwards compatibility' of some sort so I kept the `ExplorerBase` class. 

Regarding your last point, I have noticed this too - a query could be seen as a component of sorts... However it should be possible to have a generic `QueryComponent` that executes a given query in a component context (and can be depended on etc.). This might make more sense than adding the component interface to all our query definitions.

Reply:
Sorry, but it is unclear to me how the 'backwards compatibility' is useful (maybe you explained this last week, but I forgot).

The idea with the `QueryComponent` seems like a good alternative. Indeed the `ExplorerComponent` is now a class and it has much more things inside than necessary for a query. My suggestion actually requires the transformation of `QueryComponent` into an interface which has a single `Explore` method, and the other changes, where the linking between dependent components done with reflection/"dependency injection".

Reply:
The only reason I tried to keep backwards compatibility is so that we don't have to refactor all the code at once, ie. we can migrate the explorers gradually.

Reply:
Regarding the `QueryComponent`, I was thinking of something much simpler, basically a class that inherits `ExplorerComponent` and takes a `DQuery` as a constructor argument. The `Explore()` method then just executes the query and returns the result.